### PR TITLE
[FLINK-35325][cdc-connector][paimon]Support for specifying column order.

### PR DESCRIPTION
**Pull request by joyCurry30**
Refer to jira: https://issues.apache.org/jira/browse/FLINK-35325

Adding support for specifying column order when adding new columns to a table.

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/PaimonMetadataApplier.java [hunk 1]
```diff
@@ -129,25 +130,127 @@ private void applyCreateTable(CreateTableEvent event)
...
+        List<String> columnNames = table.rowType().getFieldNames();
+        String lastColumnName = columnNames.get(columnNames.size() - 1);
+        return SchemaChange.addColumn(
+                columnWithPosition.getAddColumn().getName(),
+                LogicalTypeConversion.toDataType(
+                        DataTypeUtils.toFlinkDataType(columnWithPosition.getAddColumn().getType())
+                                .getLogicalType()),
+                columnWithPosition.getAddColumn().getComment(),
+                SchemaChange.Move.after(
+                        columnWithPosition.getAddColumn().getName(), lastColumnName));
```
Comment (lvyanquan) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/PaimonMetadataApplier.java:220]: `SchemaChange.Move.after(columnWithPosition.getAddColumn().getName(), lastColumnName)` can be removed as it's the default behavior of [SchemaManager](https://github.com/apache/paimon/blob/master/paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java).

**Review by lvyanquan — COMMENTED** [review]
Thanks @joyCurry30 for this, left some minor comments.

**Review by lvyanquan — APPROVED** [review]
LGTM.

CC @PatrickRen

**Comment by yuxiqian** [comment]
Hi @joyCurry30, could you please backport it to `release-3.1` branch too?

**Comment by joyCurry30** [comment]
> Hi @joyCurry30, could you please backport it to `release-3.1` branch too?

https://github.com/apache/flink-cdc/pull/3376
cc plz. @yuxiqian

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/PaimonMetadataApplier.java [hunk 2]
```diff
@@ -129,25 +133,113 @@ private void applyCreateTable(CreateTableEvent event)
...
 
+    private List<SchemaChange> applyAddColumnEventWithPosition(AddColumnEvent event)
+            throws Catalog.TableNotExistException {
+        List<SchemaChange> tableChangeList = new ArrayList<>();
+        for (AddColumnEvent.ColumnWithPosition columnWithPosition : event.getAddedColumns()) {
+            SchemaChange tableChange;
+            switch (columnWithPosition.getPosition()) {
+                case FIRST:
+                    tableChange =
+                            SchemaChange.addColumn(
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/PaimonMetadataApplier.java:151]: What about declaring `applyAddColumnWithFirstPosition` and `applyAddColumnWithAfterPosition` too for consistency? And maybe we can extract some sharing logic (like column name, data conversion, and comments) to reduce code duplication.

**Review by yuxiqian — APPROVED** [review]

**Review by leonardBang — APPROVED** [review]
Thanks @joyCurry30 for the contribution, LGTM

**Pull request merged**